builtin/credential/aws/client.go

```diff
@@ -0,0 +1,116 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/vault/logical"
+)
+
+// getClientConfig creates a aws-sdk-go config, which is used to create client
+// that can interact with AWS API. This builds credentials in the following
+// order of preference:
+//
+// * Static credentials from 'config/client'
+// * Environment variables
+// * Instance metadata role
+func (b *backend) getClientConfig(s logical.Storage, region string) (*aws.Config, error) {
+	// Read the configured secret key and access key
+	config, err := b.clientConfigEntry(s)
+	if err != nil {
+		return nil, err
+	}
+
+	var providers []credentials.Provider
+
+	if config != nil {
+		switch {
+		case config.AccessKey != "" && config.SecretKey != "":
+			// Add the static credential provider
+			providers = append(providers, &credentials.StaticProvider{
+				Value: credentials.Value{
+					AccessKeyID:     config.AccessKey,
+					SecretAccessKey: config.SecretKey,
+				}})
+		case config.AccessKey == "" && config.AccessKey == "":
+			// Attempt to get credentials from the IAM instance role below
+		default: // Have one or the other but not both and not neither
+			return nil, fmt.Errorf(
+				"static AWS client credentials haven't been properly configured (the access key or secret key were provided but not both); configure or remove them at the 'config/client' endpoint")
+		}
+	}
+
+	// Add the environment credential provider
+	providers = append(providers, &credentials.EnvProvider{})
+
+	// Add the instance metadata role provider
+	// Create the credentials required to access the API.
+	providers = append(providers, &ec2rolecreds.EC2RoleProvider{
+		Client: ec2metadata.New(session.New(&aws.Config{
+			Region:     aws.String(region),
+			HTTPClient: cleanhttp.DefaultClient(),
+		})),
+		ExpiryWindow: 15,
+	})
+
+	creds := credentials.NewChainCredentials(providers)
+	if creds == nil {
+		return nil, fmt.Errorf("could not compile valid credential providers from static config, environemnt, or instance metadata")
+	}
+
+	// Create a config that can be used to make the API calls.
+	return &aws.Config{
```

Comment:
Hey @vishalnayak,

Sorry for being unclear! This is completely orthogonal to the credentials. At a literal level, I'm talking about having a way to configure Vault to populate the Endpoint field in the aws.Config struct (i.e., [this field](http://docs.aws.amazon.com/sdk-for-go/api/aws/Config.html#Endpoint-field)) with a custom value. So, for example, rather than send requests to https://ec2.us-east-1.amazonaws.com, I would send them to https://my-custom-ec2-endpoint.example.com

Does that make sense?

Thanks!